lib/cartage/plugins/remote.rb

```diff
@@ -0,0 +1,481 @@
+  #                   used in the +build+ script.
+  #                     bundle install --path %{remote_bundle}
+  # +dependency_cache+:: The +dependency_cachevendor_cache+ for the remote
+  #                      server. Set the same as +project_path+.
+  # +config_file+:: The remote filename of the computed Cartage configuration.
+  #                 Must be provided to the remote run of +cartage+.
+  #                   bundle exec cartage --config-file %{config_file} pack
+  # +build_script+:: The full path to the remote build script.
+  #                  <tt><em>isolation_path</em>/cartage-build-remote</tt>.
+  #
+  # == Configuration Example
+  #
+  #   ---
+  #   plugins:
+  #     remote:
+  #       hosts:
+  #         default: build-server
+  #       script: |
+  #         #! /bin/bash
+  #         bundle install --path %{remote_bundle} &&
+  #         bundle exec cartage --config-file %{config_file} pack &&
+  #         bundle exec cartage --config-file %{config_file} s3 put
+  class Remote < Cartage::Plugin
+    VERSION = '2.0.rc4' #:nodoc:
+
+    # Build on the remote server.
+    def build
+      stage.trigger! :local_setup
+      stage.trigger! :ssh_setup
+      stage.trigger! :run_prebuild
+      stage.trigger! :clone_remote
+      stage.trigger! :build_remote
+      stage.trigger! :clean_remote
+      stage.trigger! :complete
+    rescue Cartage::CLI::CustomExit
+      raise
+    rescue => e
+      error = e.exception("Remote error in stage #{stage.state}: #{e.message}")
+      error.set_backtrace(e.backtrace)
+      raise error
+    ensure
+      if postbuild_script
+        cartage.display 'Running postbuild script...'
+        system make_tmpscript('postbuild', postbuild_script, subs).path,
+          stage.state.to_s, error.to_s
+      end
+
+      tmpfiles.each do |tmpfile|
+        tmpfile.close
+        tmpfile.unlink
+      end
+      tmpfiles.clear
+    end
+
+    # Check that the configuration is correct. If +require_host+ is present, an
+    # exception will be thrown if a host is required and not present.
+    def check_config(require_host: false)
+      hosts = cartage.config(for_plugin: :remote).hosts
+      verify_hosts(hosts)
+      fail "No host #{name} present" if require_host && !hosts.dig(name)
+    end
+
+    private
+
+    def initialize(*) #:nodoc:
+      super
+      @host = @name = @keys = @key_data = nil
+
+      @stage = MicroMachine.new('new').tap { |stage|
+        stage.when :local_setup, 'new' => 'local_config'
+        stage.when :ssh_setup, 'local_config' => 'ssh_config'
+        stage.when :run_prebuild, 'ssh_config' => 'prebuild'
+        stage.when :clone_remote, 'prebuild' => 'remote_clone'
+        stage.when :build_remote, 'remote_clone' => 'remote_build'
+        stage.when :clean_remote, 'remote_build' => 'cleanup'
+        stage.when :complete, 'cleanup' => 'finished'
+        stage.on(:any) { |event| dispatch(event, stage.state) }
+      }
```

Comment:
This is coooool!
[bitmoji](https://render.bitstrips.com/v2/cpanel/8643870-125376252_1-s1-v1.png?palette=1&width=246)